File: www/src/systems/collision.ts
class CollisionSystem extends System {

    update(_stateManager: StateManager, entityManager: EntityManager, delta: number, tick: number) {
        let playerComponents = entityManager.getComponents(
            true,
            CONSTANTS.COMPONENT.VELOCITY, 
            CONSTANTS.COMPONENT.TRANSFORM,
            CONSTANTS.COMPONENT.JUMPING);
        
        let ballComponent = entityManager.getComponents(
            true,
            CONSTANTS.COMPONENT.VELOCITY, 
            CONSTANTS.COMPONENT.TRANSFORM,
            CONSTANTS.COMPONENT.BALL);

        Object.keys(playerComponents)
            .forEach(function(eid){

                Object.keys(ballComponent)
                    .forEach(function(bid){

                        let dx = playerComponents[eid].transform.x - ballComponent[bid].transform.x;
                        let dy = playerComponents[eid].transform.y - ballComponent[bid].transform.y;

                        // Distance from ball to player
                        let dist = Math.sqrt(dx*dx + dy*dy);
                        
                        if(dist < (50 + 5 + 6)){


                            let a = Math.atan(dy/dx) + (Math.sign(dx) === 1 ? -Math.PI : 0);

                            console.log("TICK " + tick + " HIT!" + " DX: " + dx + ", DY:" + dy + ", VY: " + playerComponents[eid].velocity.y + ", DIST: " + dist + " ANGLE1: " + a + ", X:" + playerComponents[eid].transform.x + ", Y: " + playerComponents[eid].transform.y + ", X1:" + ballComponent[bid].transform.x + ", Y1:" + ballComponent[bid].transform.y);

                            ballComponent[bid].transform.y = Math.round(playerComponents[eid].transform.y + (50 + 5 + 6) * Math.sin(a));
                            ballComponent[bid].transform.x = Math.round(playerComponents[eid].transform.x + (50 + 5 + 6) * Math.cos(a));
                            ballComponent[bid].velocity.y = Math.min(15, (10 + playerComponents[eid].velocity.y));
                            ballComponent[bid].velocity.x = dx * -1;
                        }

                        
                    });

            });
        
        Object.keys(ballComponent)
            .forEach(function(bid){

                if(ballComponent[bid].transform.x > 1200){
                    ballComponent[bid].transform.x = 1200;

                    console.log("RIGHT WALL HIT!");

                    if(ballComponent[bid].velocity.x > 0){
                        ballComponent[bid].velocity.x = ballComponent[bid].velocity.x * -1;
                    }
                }else if(ballComponent[bid].transform.x < 0){
                    ballComponent[bid].transform.x = 0;

                    
                    console.log("LEFT WALL HIT!");

                    if(ballComponent[bid].velocity.x < 0){
                        ballComponent[bid].velocity.x = ballComponent[bid].velocity.x * -1;
                    }
                
                // ROPE COLLISION
                }else if(ballComponent[bid].transform.x > 595 && 
                        ballComponent[bid].transform.x < 605 && 
                        ballComponent[bid].transform.y < 150){

                    
                    console.log("ROPE HIT!");
                    ballComponent[bid].velocity.x = ballComponent[bid].velocity.x * -1;
                }

                // FLOOR COLLISION
                if(ballComponent[bid].transform.y <= 0){
                    ballComponent[bid].transform.y = 400;
                    ballComponent[bid].velocity.x = 0;
                    ballComponent[bid].velocity.y = 0;

                }
            });
    }
}

File: internal/app/server/system/collision.go
package system

import (
	"fmt"
	"math"

	"stonecastle.internal.stonepath.se/pgstenberg/volleyball/internal/app/server/component"
	"stonecastle.internal.stonepath.se/pgstenberg/volleyball/internal/app/server/constant"
	"stonecastle.internal.stonepath.se/pgstenberg/volleyball/internal/pkg/core"
)

type CollisionSystem struct {
}

func (cs *CollisionSystem) Update(entityManager *core.EntityManager, tick uint16, pause bool, delta float64) {

	if pause {
		return
	}

	playerComponent := entityManager.GetComponents(true, constant.VelocityComponent, constant.TransformComponent, constant.PlayerComponent)
	ballComponent := entityManager.GetComponents(true, constant.VelocityComponent, constant.TransformComponent, constant.BallComponent, constant.NetworkComponent)

	/*
		Player with Ball collision stuff
	*/
	for _, pcomponents := range playerComponent {

		pvc := (*pcomponents[constant.VelocityComponent]).(*component.VelocityComponent)
		ptc := (*pcomponents[constant.TransformComponent]).(*component.TransformComponent)

		for beid, bcomponents := range ballComponent {

			bvc := (*bcomponents[constant.VelocityComponent]).(*component.VelocityComponent)
			btc := (*bcomponents[constant.TransformComponent]).(*component.TransformComponent)
			bnc := (*bcomponents[constant.NetworkComponent]).(*component.NetworkComponent)

			bnc.RequireSynchronize = false

			dx := float64(int(ptc.PositionX) - int(btc.PositionX))
			dy := float64(int(ptc.PositionY) - int(btc.PositionY))

			dist := math.Sqrt(dx*dx + dy*dy)

			if dist < (50 + 5 + 6) {
				a0 := float64(0)
				if dx > 0 {
					a0 = -math.Pi
				}
				a := math.Atan(dy/dx) + a0

				fmt.Printf("%d %s    HIT    DX: %04d    DY: %04d    DELTA: %04f\n", tick, beid, dx, dy, dist)

				bnc.RequireSynchronize = true

				btc.PositionY = uint16(math.Round(float64(ptc.PositionY) + (50+5+6)*math.Sin(a)))
				btc.PositionX = uint16(math.Round(float64(ptc.PositionX) + (50+5+6)*math.Cos(a)))
				bvc.VelocityY = math.Min(15, (10 + pvc.VelocityY))
				bvc.VelocityX = float64(dx) * -1
			}
		}

	}

	/*
		Ball Collision Stuff
	*/
	for beid, components := range ballComponent {

		vc := (*components[constant.VelocityComponent]).(*component.VelocityComponent)
		tc := (*components[constant.TransformComponent]).(*component.TransformComponent)
		bnc := (*components[constant.NetworkComponent]).(*component.NetworkComponent)

		if tc.PositionX > 1200 {
			tc.PositionX = 1200

			fmt.Printf("%d %s    HIT-RIGHT-WALL    X: %04d    Y: %04d\n", tick, beid, tc.PositionX, tc.PositionY)

			bnc.RequireSynchronize = true

			if vc.VelocityX > 0 {
				vc.VelocityX = vc.VelocityX * -1
			}
		} else if tc.PositionX < 0 {
			tc.PositionX = 0

			fmt.Printf("%d %s    HIT-LEFT-WALL    X: %04d    Y: %04d\n", tick, beid, tc.PositionX, tc.PositionY)

			bnc.RequireSynchronize = true

			if vc.VelocityX < 0 {
				vc.VelocityX = vc.VelocityX * -1
			}
			// ROPE COLLISION
		} else if tc.PositionX > 595 &&
			tc.PositionX < 605 &&
			tc.PositionY < 150 {

			fmt.Printf("%d %s    HIT-ROPE    X: %04d    Y: %04d\n", tick, beid, tc.PositionX, tc.PositionY)

			vc.VelocityX = vc.VelocityX * -1
		} else if tc.PositionY <= 50 {
			tc.PositionY = 50
			vc.VelocityX = 0
			vc.VelocityY = 0
		}

	}
}
